src/icf/utils/util.ts

```diff
@@ -0,0 +1,155 @@
+        .map((subType) => transformType(subType, typeToStruct, structToAdd))
+        .join(',')})`;
+      return ret;
+    }
+    const structName = `struct_${hashType(type)}`;
+    if (structToAdd !== undefined) {
+      structToAdd.set(type, {
+        name: structName,
+        type: 'struct',
+        members: subTypes.map((subType, index) => ({
+          name: `member_${index}`,
+          type: transformType(subType, typeToStruct, structToAdd),
+        })),
+      });
+    }
+    return structName;
+  }
+  return type;
+}
+
+export function castStatement(
+  lvar: string,
+  rvar: string,
+  type: string,
+  typeToStruct: Map<string, StructAbiItemType>,
+  addedStruct?: Map<string, StructAbiItemType>,
+): string {
+  type = type.trim();
+  if (type === 'felt') return `${INDENT}let (${lvar}) = narrow_safe(${rvar});`;
+  if (typeToStruct.has(type)) return `${INDENT}let (${lvar}) = ${type}_cast(${rvar});`;
+  if (type.endsWith('*')) {
+    return `${INDENT}let ${lvar} = cast(${rvar}, ${type});`;
+  }
+  if (type.startsWith('(') && type.endsWith(')')) {
+    const subTypes = tupleParser(type);
+    const castBody = [];
+    for (let i = 0; i < subTypes.length; i++) {
+      castBody.push(
+        castStatement(
+          `${lvar}_${i}`,
+          addedStruct?.has(type) ? `${rvar}.member_${i}` : `${rvar}[${i}]`,
+          subTypes[i],
+          typeToStruct,
+          addedStruct,
+        ),
+      );
+    }
+    castBody.push(`${INDENT}let ${lvar} = (${subTypes.map((_, i) => `${lvar}_${i}`).join(',')});`);
+    return castBody.join('\n');
+  }
+  return `${INDENT}let ${lvar} = ${rvar};`;
+}
+
+export function reverseCastStatement(
+  lvar: string,
+  rvar: string,
+  type: string,
+  typeToStruct: Map<string, StructAbiItemType>,
+  addedStruct?: Map<string, StructAbiItemType>,
+): string {
+  type = type.trim();
+  if (type === 'felt') return `${INDENT}let (${lvar}) = felt_to_uint256(${rvar});`;
+  if (typeToStruct.has(type)) return `${INDENT}let (${lvar}) = ${type}_cast_reverse(${rvar});`;
+  if (type.endsWith('*')) {
+    return `${INDENT}let ${lvar} = cast(${rvar}, ${transformType(type, typeToStruct)});`;
+  }
+  if (type.startsWith('(') && type.endsWith(')')) {
+    const subTypes = tupleParser(type);
+    const castBody = [];
+    for (let i = 0; i < subTypes.length; i++) {
+      castBody.push(
+        reverseCastStatement(
+          `${lvar}_${i}`,
+          `${rvar}[${i}]`,
+          subTypes[i],
+          typeToStruct,
+          addedStruct,
+        ),
+      );
+    }
+    castBody.push(
+      `${INDENT}let ${lvar} = ${addedStruct?.has(type) ? `struct_${hashType(type)}` : ``}(${subTypes
+        .map((_, i) => `${lvar}_${i}`)
+        .join(',')});`,
+    );
+    return castBody.join('\n');
+  }
+  return `${INDENT}let ${lvar} = ${rvar};`;
+}
+
+export function tupleParser(tuple: string): string[] {
+  if (tuple.startsWith('(') && tuple.endsWith(')')) tuple = tuple.slice(1, -1).trim();
```

Comment:
question: what happens for the type : `((felt, felt), (felt,felt)`

Reply:
are we not expecting nested tuples, what about static 2d arrays

Reply:
It'll support nested tuples, line:132 is just to remove the enclosing parenthesis

Reply:
And, this function doesn't expect faulty tuple type strings. I'll add a case when it receives a wrong string